src/ops/shapes.rs

```diff
@@ -0,0 +1,2528 @@
+        if low < 0 || high < 0 {
+            return Err(PadError::NegativeWidth {
+                axis,
+                size: (low, high),
+            });
+        }
+    }
+
+    unsafe {
+        let c_array = mlx_sys::mlx_pad(
+            array.c_array,
+            axes.as_ptr(),
+            axes.len(),
+            low_pads.as_ptr(),
+            low_pads.len(),
+            high_pads.as_ptr(),
+            high_pads.len(),
+            value.c_array,
+            stream.as_ptr(),
+        );
+        Ok(Array::from_ptr(c_array))
+    }
+}
+
+/// Pad an array with a constant value. Panics if the width is invalid.
+///
+/// # Params
+///
+/// - `array`: The input array.
+/// - `width`: Number of padded values to add to the edges of each axis:`((before_1, after_1),
+///   (before_2, after_2), ..., (before_N, after_N))`. If a single pair of integers is passed then
+///   `(before_i, after_i)` are all the same. If a single integer or tuple with a single integer is
+///   passed then all axes are extended by the same number on each side.
+/// - `value`: The value to pad the array with. Default is `0` if not provided.
+///
+/// # Panics
+///
+/// Panics if the width is invalid. See [`try_pad`] for more information.
+///
+/// # Example
+///
+/// ```rust
+/// use mlx::{prelude::*, ops::*};
+///
+/// let a = Array::from_iter(0..4, &[2, 2]);
+/// let result = pad(&a, 1, Array::from_int(0));
+/// ```
+#[default_device]
+pub fn pad_device<'a>(
+    array: &'a Array,
+    width: impl Into<PadWidth<'a>>,
+    value: impl Into<Option<Array>>,
+    stream: StreamOrDevice,
+) -> Array {
+    try_pad_device(array, width, value, stream).unwrap()
+}
+
+fn stack_inner(arrays: &[impl AsRef<Array>], axis: i32, stream: StreamOrDevice) -> Array {
+    unsafe {
+        let c_arrays = new_mlx_vector_array(arrays);
+        let c_array = mlx_sys::mlx_stack(c_arrays, axis, stream.as_ptr());
+
+        let result = Array::from_ptr(c_array);
+        mlx_sys::mlx_free(c_arrays as *mut c_void);
+
+        result
+    }
+}
+
+/// Stacks the arrays along a new axis.
+///
+/// # Params
+///
+/// - `arrays`: The input arrays.
+/// - `axis`: The axis in the result array along which the input arrays are stacked.
+///
+/// # Safety
+///
+/// The function is unsafe because it does not check if the arrays have the same shape.
+///
+/// # Example
+///
+/// ```rust
+/// use mlx::{prelude::*, ops::*};
+///
+/// let a = Array::from_iter(0..4, &[2, 2]);
+/// let b = Array::from_iter(4..8, &[2, 2]);
+/// let result = unsafe { stack_unchecked(&[&a, &b], 0) };
+/// ```
+#[default_device]
+pub unsafe fn stack_device_unchecked(
+    arrays: impl IntoIterator<Item = impl AsRef<Array>>,
+    axis: i32,
+    stream: StreamOrDevice,
+) -> Array {
+    let arrays = arrays.into_iter().collect::<Vec<_>>();
+    stack_inner(&arrays, axis, stream)
+}
+
+#[default_device]
+pub fn try_stack_device(
+    arrays: impl IntoIterator<Item = impl AsRef<Array>>,
```

Comment:
Why does this have to be `IntoIterator`? Seems like ` &[impl AsRef<Array>],` would work just as well

Reply:
True. Not sure why I didn't see this lol

Reply:
Nvm, there are places where directly accessing the first element (multiple times) is needed

Reply:
Addressed in ec661bf335722cc03b7c0da398d6e683b0810bc1